include/oneapi/tbb/collaborative_call_once.h

```diff
@@ -0,0 +1,222 @@
+    #pragma warning (push)
+    #pragma warning (disable: 4324)
+#endif
+
+constexpr std::size_t bit_count = 7;
+
+static_assert(1 << bit_count == max_nfs_size, "bit_count must be a log2(max_nfs_size)");
+
+inline constexpr std::uintptr_t maskoff_pointer(std::uintptr_t ptr) {
+    return ptr & (std::size_t(-1) << bit_count);
+}
+
+class alignas(max_nfs_size) once_runner {
+
+    struct storage_t {
+        task_arena my_arena{ task_arena::attach{} };
+        wait_context my_wait_context{0};
+        task_group_context my_context{ task_group_context::bound,
+            task_group_context::default_traits | task_group_context::concurrent_wait };
+    };
+
+    template <typename Func>
+    class call_once_delegate : public delegate_base {
+    public:
+        call_once_delegate(Func& f) : my_func(f) {}
+
+        bool operator()() const override {
+            my_func();
+            return true;
+        }
+
+    private:
+        Func& my_func;
+    };
+
+    std::atomic<std::int64_t> ref_count{0};
+    std::atomic<bool> m_initialized{false};
+
+    // Storage with task_arena and contexts must be initialized only by winner thread
+    // by calling make_runner() method
+    union {
+        storage_t m_storage;
+    };
+
+    template<typename Fn>
+    void isolated_execute(Fn f) {
+        call_once_delegate<Fn> delegate(f);
+
+        r1::isolate_within_arena(delegate, reinterpret_cast<std::intptr_t>(this));
+    }
+
+public:
+    once_runner() {}
+
+    ~once_runner() {
+        spin_wait_while(ref_count, [&](std::int64_t value) { return value > 0; }, std::memory_order_acquire);
+        if (m_initialized.load(std::memory_order_relaxed)) {
+            m_storage.~storage_t();
+        }
+    }
+
+    void make_runner() {
+        new(&m_storage) storage_t();
+        m_initialized.store(true, std::memory_order_release);
+    }
+
+    void wait_for_init() {
+        spin_wait_while_eq(m_initialized, false);
+    }
+
+    void increase_ref() { ref_count++; }
+
+    void decrease_ref() { ref_count--; }
+
+    template <typename F>
+    void run_once(F&& f) {
+        m_storage.my_arena.execute([&] {
+            isolated_execute([&] {
+                function_stack_task<F> t{ std::forward<F>(f), m_storage.my_wait_context };
+                m_storage.my_wait_context.reserve();
+
+                execute_and_wait(t, m_storage.my_context, m_storage.my_wait_context, m_storage.my_context);
+            });
+        });
+    }
+
+    void assist() noexcept {
+        m_storage.my_arena.execute([&] {
+            isolated_execute([&] {
+                // We do not want to get an exception from user functor on moonlighting threads.
+                // The exception is handled with the winner thread
+                task_group_context stub_context;
+                wait(m_storage.my_wait_context, stub_context);
+            });
+        });
+    }
+
+};
+
+class runner_lifetime_checker : no_copy {
```

Comment:
Make it a nested class in `once_runner` and make private `ref` manipulation functions. 

`runner_lifetime_checker`->`once_runner::lifetime_tracker`